ipfs/src/lib.rs

```diff
@@ -23,31 +23,113 @@ extern crate cid;
 extern crate rlp;
 extern crate ethcore;
 extern crate ethcore_util as util;
+extern crate jsonrpc_http_server;
 
-mod error;
-mod handler;
+pub mod error;
+mod route;
 
 use std::io::Write;
 use std::sync::Arc;
-use std::net::{SocketAddr, IpAddr, Ipv4Addr};
+use std::net::{SocketAddr, IpAddr};
 use error::ServerError;
-use handler::{IpfsHandler, Out};
+use route::Out;
+use jsonrpc_http_server::cors;
 use hyper::server::{Listening, Handler, Request, Response};
 use hyper::net::HttpStream;
-use hyper::header::{ContentLength, ContentType, Origin};
+use hyper::header::{ContentLength, ContentType, AccessControlAllowOrigin};
 use hyper::{Next, Encoder, Decoder, Method, RequestUri, StatusCode};
 use ethcore::client::BlockChainClient;
 
+
+/// Request/response handler
+pub struct IpfsHandler {
+	/// Response to send out
+	out: Out,
+
+	/// How many bytes from the response have been written
+	out_progress: usize,
+
+	/// Origin request header
+	origin: Option<String>,
+
+	/// Allowed CORS domains
+	cors_domains: Option<Vec<AccessControlAllowOrigin>>,
+
+	/// Hostnames allowed in the `Host` request header
+	allowed_hosts: Option<Vec<String>>,
+
+	/// Reference to the Blockchain Client
+	client: Arc<BlockChainClient>,
+}
+
+impl IpfsHandler {
+	pub fn client(&self) -> &BlockChainClient {
+		&*self.client
+	}
+
+	pub fn new(cors: Option<Vec<String>>, hosts: Option<Vec<String>>, client: Arc<BlockChainClient>) -> Self {
+		fn origin_to_header(origin: String) -> AccessControlAllowOrigin {
+			if origin == "*" {
+				return AccessControlAllowOrigin::Any;
+			}
+
+			AccessControlAllowOrigin::Value(origin)
+		}
+
+		IpfsHandler {
+			out: Out::Bad("Invalid Request"),
+			out_progress: 0,
+			origin: None,
+			cors_domains: cors.map(|vec| vec.into_iter().map(origin_to_header).collect()),
+			allowed_hosts: hosts,
+			client: client,
+		}
+	}
+
+	fn is_host_allowed(&self, req: &Request<HttpStream>) -> bool {
+		match self.allowed_hosts {
+			Some(ref hosts) => jsonrpc_http_server::is_host_header_valid(&req, hosts),
+			None => true,
+		}
+	}
+
+	fn is_origin_allowed(&self) -> bool {
+		// Check origin header first, no header passed is good news
+		let origin = match self.origin {
+			Some(ref origin) => origin,
+			None => return true,
+		};
+
+		let cors_domains = match self.cors_domains {
+			Some(ref domains) => domains,
+			None => return false,
+		};
+
+		cors_domains.iter().any(|domain| match *domain {
+			AccessControlAllowOrigin::Value(ref allowed) => origin == allowed,
+			AccessControlAllowOrigin::Any => true,
```

Comment:
This is already fixed, you just can't see it on the diff here (it's one line below).